dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-fee-record-payment-entity-groups-to-fee-record-payment-groups.ts

```diff
@@ -1,56 +1,123 @@
+export const mapFeeRecordPaymentEntityGroupsToFeeRecordPaymentGroups = async (
   feeRecordPaymentEntityGroups: FeeRecordPaymentEntityGroup[],
-): FeeRecordPaymentGroup[] => {
-  return feeRecordPaymentEntityGroups.map(({ feeRecords: feeRecordEntitiesInGroup, payments }) => {
-    const status = getStatusForGroupOfFeeRecords(feeRecordEntitiesInGroup);
+): Promise<FeeRecordPaymentGroup[]> =>
+  await Promise.all(
+    feeRecordPaymentEntityGroups.map(async ({ feeRecords: feeRecordEntitiesInGroup, payments }) => {
+      const groupStatus = getStatusForGroupOfFeeRecords(feeRecordEntitiesInGroup);
+      const { reconciledByUser, dateReconciled } = await getReconciliationDataForFeeRecords(feeRecordEntitiesInGroup, groupStatus);
 
-    if (payments.length === 0) {
-      // If there are no payments, there is only one fee record in the group
-      const feeRecord = mapFeeRecordEntityToFeeRecord(feeRecordEntitiesInGroup[0]);
-      const totalReportedPayments = feeRecord.reportedPayments;
+      if (payments.length === 0) {
+        // If there are no payments, there is only one fee record in the group
+        const feeRecord = mapFeeRecordEntityToFeeRecord(feeRecordEntitiesInGroup[0]);
+        const totalReportedPayments = feeRecord.reportedPayments;
 
-      return {
-        feeRecords: [feeRecord],
-        totalReportedPayments,
-        paymentsReceived: null,
-        totalPaymentsReceived: null,
-        status,
-      };
-    }
+        return {
+          feeRecords: [feeRecord],
+          totalReportedPayments,
+          paymentsReceived: null,
+          totalPaymentsReceived: null,
+          status: groupStatus,
+          reconciledByUser,
+          dateReconciled,
+        };
+      }
 
-    const feeRecords = feeRecordEntitiesInGroup.map(mapFeeRecordEntityToFeeRecord);
+      const feeRecords = feeRecordEntitiesInGroup.map(mapFeeRecordEntityToFeeRecord);
 
-    const allReportedPayments = feeRecords.map(({ reportedPayments }) => reportedPayments);
-    const totalReportedPayments = calculateTotalCurrencyAndAmount(allReportedPayments);
+      const allReportedPayments = feeRecords.map(({ reportedPayments }) => reportedPayments);
+      const totalReportedPayments = calculateTotalCurrencyAndAmount(allReportedPayments);
 
-    const paymentsReceived = payments.map(mapPaymentEntityToPayment);
-    const totalPaymentsReceived = calculateTotalCurrencyAndAmount(paymentsReceived);
+      const paymentsReceived = payments.map(mapPaymentEntityToPayment);
+      const totalPaymentsReceived = calculateTotalCurrencyAndAmount(paymentsReceived);
 
-    return {
-      feeRecords,
-      totalReportedPayments,
-      paymentsReceived,
-      totalPaymentsReceived,
-      status,
-    };
-  });
-};
+      return {
+        feeRecords,
+        totalReportedPayments,
+        paymentsReceived,
+        totalPaymentsReceived,
+        status: groupStatus,
+        reconciledByUser,
+        dateReconciled,
+      };
```

Comment:
We could still pull a method for the interior of the map but not make it an exported member and then it is just called by this function. It would reduce the nesting slightly so might be nice?

Reply:
Pulling the logic out into an individual mapping function means the user of the function needs to remember to use `await Promise.all`, but I'd prefer to keep that logic contained in this file. As all the unit tests and existing implementation is built to work with this version, I'd rather keep it